iobuf: Call iobuf_get_from_small to allocate buffer by iobuf_get_from…

mohit84 opened the pull request:
…_stdalloc

The function iobuf_get_from_stdalloc call while a buffer_size is
greater than 1M and the function iobuf_get_from_small call
while size is less than 128k the only difference is iobuf_get_from_small
does not aligned the buffer so call iobuf_get_from_small to allocate
a buffer and realigned the buffer separately

Fixes: #2850
Signed-off-by: Mohit Agrawal <moagrawa@redhat.com>

mohit84 committed:
iobuf: Call iobuf_get_from_small to allocate buffer by iobuf_get_from_stdalloc

The function iobuf_get_from_stdalloc call while a buffer_size is
greater than 1M and the function iobuf_get_from_small call
while size is less than 128k the only difference is iobuf_get_from_small
does not aligned the buffer so call iobuf_get_from_small to allocate
a buffer and realigned the buffer separately

Fixes: #2850
Signed-off-by: Mohit Agrawal <moagrawa@redhat.com>

mohit84 commented:
/run regression

mykaul commented on libglusterfs/src/iobuf.c:
```diff
@@ -468,35 +468,24 @@ __iobuf_get(struct iobuf_pool *iobuf_pool, const size_t page_size,
     struct iobuf *iobuf = NULL;
-    struct iobuf_arena *iobuf_arena = NULL;
-    struct iobuf_arena *trav = NULL;
     int ret = -1;
```
nit: unrelated to this patch, I'm not sure if ret is really needed here. I would have done a 'return iobuf' instead of 'ret = 0' later on. (but again, has nothing to do with this change)

mykaul commented:
I have a different idea - check the size you need to allocate. If it's large, align ptr. If not, don't. I think we could easily get rid of the whole fake arena this way.

gluster-ant commented:
1 test(s) failed 
./tests/basic/afr/granular-esh/replace-brick.t

0 test(s) generated core 


4 test(s) needed retry 
./tests/000-flaky/basic_mount-nfs-auth.t
./tests/000-flaky/bugs_nfs_bug-1116503.t
./tests/000-flaky/glusterd-restart-shd-mux.t
./tests/basic/afr/granular-esh/replace-brick.t

2 flaky test(s) marked as success even though they failed 
./tests/000-flaky/basic_mount-nfs-auth.t
./tests/000-flaky/bugs_nfs_bug-1116503.t
https://build.gluster.org/job/gh_centos7-regression/1693/

mohit84 committed:
iobuf: Resolve reviewer comments

Use single iobuf_get_from_stdalloc function to allocate memory

Fixes: #2850
Signed-off-by: Mohit Agrawal <moagrawa@redhat.com>

mohit84 commented:
> I have a different idea - check the size you need to allocate. If it's large, align ptr. If not, don't. I think we could easily get rid of the whole fake arena this way.

done

mohit84 commented:
/run regression

mykaul commented:
If you got rid of the fake arena, I think this line:
```
    for (i = 0; i <= IOBUF_ARENA_MAX_INDEX; i++) {
```
should be:
```
    for (i = 0; i < IOBUF_ARENA_MAX_INDEX; i++) {
```

(in iobuf_pool_new() code)

mohit84 commented:
> If you got rid of the fake arena, I think this line:
> 
> ```
>     for (i = 0; i <= IOBUF_ARENA_MAX_INDEX; i++) {
> ```
> 
> should be:
> 
> ```
>     for (i = 0; i < IOBUF_ARENA_MAX_INDEX; i++) {
> ```
> 
> (in iobuf_pool_new() code)

done

mohit84 commented:
/run regression

mohit84 commented:
/run regression

mohit84 commented:
/run regression

mohit84 commented:
/run full regression

gluster-ant commented:
1 test(s) failed 
./tests/00-geo-rep/00-georep-verify-non-root-setup.t

0 test(s) generated core 


15 test(s) needed retry 
./tests/000-flaky/basic_afr_split-brain-favorite-child-policy.t
./tests/000-flaky/basic_changelog_changelog-snapshot.t
./tests/000-flaky/basic_distribute_rebal-all-nodes-migrate.t
./tests/000-flaky/basic_ec_ec-quorum-count-partial-failure.t
./tests/000-flaky/basic_mount-nfs-auth.t
./tests/000-flaky/bugs_core_multiplex-limit-issue-151.t
./tests/000-flaky/bugs_distribute_bug-1117851.t
./tests/000-flaky/bugs_distribute_bug-1122443.t
./tests/000-flaky/bugs_glusterd_bug-857330/normal.t
./tests/000-flaky/bugs_glusterd_bug-857330/xml.t
./tests/000-flaky/bugs_glusterd_quorum-value-check.t
./tests/000-flaky/bugs_nfs_bug-1116503.t
./tests/000-flaky/features_lock-migration_lkmigration-set-option.t
./tests/000-flaky/glusterd-restart-shd-mux.t
./tests/00-geo-rep/00-georep-verify-non-root-setup.t

14 flaky test(s) marked as success even though they failed 
./tests/000-flaky/basic_afr_split-brain-favorite-child-policy.t
./tests/000-flaky/basic_changelog_changelog-snapshot.t
./tests/000-flaky/basic_distribute_rebal-all-nodes-migrate.t
./tests/000-flaky/basic_ec_ec-quorum-count-partial-failure.t
./tests/000-flaky/basic_mount-nfs-auth.t
./tests/000-flaky/bugs_core_multiplex-limit-issue-151.t
./tests/000-flaky/bugs_distribute_bug-1117851.t
./tests/000-flaky/bugs_distribute_bug-1122443.t
./tests/000-flaky/bugs_glusterd_bug-857330/normal.t
./tests/000-flaky/bugs_glusterd_bug-857330/xml.t
./tests/000-flaky/bugs_glusterd_quorum-value-check.t
./tests/000-flaky/bugs_nfs_bug-1116503.t
./tests/000-flaky/features_lock-migration_lkmigration-set-option.t
./tests/000-flaky/glusterd-restart-shd-mux.t
https://build.gluster.org/job/gh_centos7-regression/1700/

mohit84 commented:
/run regression

gluster-ant commented:
1 test(s) failed 
./tests/00-geo-rep/00-georep-verify-non-root-setup.t

0 test(s) generated core 


15 test(s) needed retry 
./tests/000-flaky/basic_afr_split-brain-favorite-child-policy.t
./tests/000-flaky/basic_changelog_changelog-snapshot.t
./tests/000-flaky/basic_distribute_rebal-all-nodes-migrate.t
./tests/000-flaky/basic_ec_ec-quorum-count-partial-failure.t
./tests/000-flaky/basic_mount-nfs-auth.t
./tests/000-flaky/bugs_core_multiplex-limit-issue-151.t
./tests/000-flaky/bugs_distribute_bug-1117851.t
./tests/000-flaky/bugs_distribute_bug-1122443.t
./tests/000-flaky/bugs_glusterd_bug-857330/normal.t
./tests/000-flaky/bugs_glusterd_bug-857330/xml.t
./tests/000-flaky/bugs_glusterd_quorum-value-check.t
./tests/000-flaky/bugs_nfs_bug-1116503.t
./tests/000-flaky/features_lock-migration_lkmigration-set-option.t
./tests/000-flaky/glusterd-restart-shd-mux.t
./tests/00-geo-rep/00-georep-verify-non-root-setup.t

14 flaky test(s) marked as success even though they failed 
./tests/000-flaky/basic_afr_split-brain-favorite-child-policy.t
./tests/000-flaky/basic_changelog_changelog-snapshot.t
./tests/000-flaky/basic_distribute_rebal-all-nodes-migrate.t
./tests/000-flaky/basic_ec_ec-quorum-count-partial-failure.t
./tests/000-flaky/basic_mount-nfs-auth.t
./tests/000-flaky/bugs_core_multiplex-limit-issue-151.t
./tests/000-flaky/bugs_distribute_bug-1117851.t
./tests/000-flaky/bugs_distribute_bug-1122443.t
./tests/000-flaky/bugs_glusterd_bug-857330/normal.t
./tests/000-flaky/bugs_glusterd_bug-857330/xml.t
./tests/000-flaky/bugs_glusterd_quorum-value-check.t
./tests/000-flaky/bugs_nfs_bug-1116503.t
./tests/000-flaky/features_lock-migration_lkmigration-set-option.t
./tests/000-flaky/glusterd-restart-shd-mux.t
https://build.gluster.org/job/gh_centos7-regression/1703/

mohit84 commented:
/run regression

mykaul commented on libglusterfs/src/iobuf.c:
```diff
@@ -561,7 +520,7 @@ iobuf_get2(struct iobuf_pool *iobuf_pool, size_t page_size)
     */
     if (page_size <= USE_IOBUF_POOL_IF_SIZE_GREATER_THAN) {
-        iobuf = iobuf_get_from_small(page_size);
+        iobuf = iobuf_get_from_stdalloc(page_size, _gf_false);
```
I'd argue that anything 4K and above we should align, but could be done in a different patch.

mohit84 commented:
/run regression

gluster-ant commented:
CLANG-FORMAT FAILURE:
Before merging the patch, this diff needs to be considered for passing clang-format

```diff --git a/libglusterfs/src/iobuf.c b/libglusterfs/src/iobuf.c
index a6fd3eca1..b7f7f64e2 100644
--- a/libglusterfs/src/iobuf.c
+++ b/libglusterfs/src/iobuf.c
@@ -304,7 +304,7 @@ iobuf_create_stdalloc_arena(struct iobuf_pool *iobuf_pool)
     iobuf_arena->page_size = 0x7fffffff;
 
     list_add_tail(&iobuf_arena->list,
-                  &iobuf_pool->arenas[IOBUF_ARENA_MAX_INDEX-1]);
+                  &iobuf_pool->arenas[IOBUF_ARENA_MAX_INDEX - 1]);
 
 err:
     return;
```

mykaul commented:
Why do we even need iobuf_create_stdalloc_arena()? we are allocating fine without this arena.

xhernandez commented on libglusterfs/src/iobuf.c:
```diff
@@ -304,7 +304,7 @@ iobuf_create_stdalloc_arena(struct iobuf_pool *iobuf_pool)
 
     list_add_tail(&iobuf_arena->list,
-                  &iobuf_pool->arenas[IOBUF_ARENA_MAX_INDEX]);
+                  &iobuf_pool->arenas[IOBUF_ARENA_MAX_INDEX-1]);
```
Why is needed this change ?

mohit84 replied:
The change is needed because iobuf_pool_new is allocating list like below 

for (i = 0; i < IOBUF_ARENA_MAX_INDEX; i++) {
        INIT_LIST_HEAD(&iobuf_pool->arenas[i]);
        INIT_LIST_HEAD(&iobuf_pool->filled[i]);
        INIT_LIST_HEAD(&iobuf_pool->purge[i]);
   }
and here when the code is trying to access IOBUF_ARENA_MAX_INDEX the process is crash.

xhernandez replied:
It's doing so because you have also changed that part of the code. But previously the last entry in the `arenas` was used. I don't see why it needs to be removed with this patch.

mohit84 replied:
We don't need to create arena in case of big buffer(>1M) or less than 128K so we don;t need to call this function.
I have upload a new patch.

mohit84 committed:
iobuf: Resolve reviewer comments

1) Don't need to create an arena for bigger buffer(>1M) because
   big buffer always allocated directly from stdalloc.
2) Remove arenas those are smaller than 128k

Fixes: #2850
Signed-off-by: Mohit Agrawal <moagrawa@redhat.com>

mohit84 commented:
/run regression

gluster-ant commented:
1 test(s) failed 
./tests/bugs/gfapi/bug-1319374-THIS-crash.t

0 test(s) generated core 


4 test(s) needed retry 
./tests/000-flaky/basic_mount-nfs-auth.t
./tests/000-flaky/bugs_nfs_bug-1116503.t
./tests/000-flaky/glusterd-restart-shd-mux.t
./tests/bugs/gfapi/bug-1319374-THIS-crash.t

2 flaky test(s) marked as success even though they failed 
./tests/000-flaky/basic_mount-nfs-auth.t
./tests/000-flaky/bugs_nfs_bug-1116503.t
https://build.gluster.org/job/gh_centos7-regression/1734/

mykaul commented on libglusterfs/src/iobuf.c:
```diff
@@ -468,79 +441,38 @@ __iobuf_get(struct iobuf_pool *iobuf_pool, const size_t page_size,
-    iobuf->iobuf_arena = iobuf_arena;
-    iobuf->page_size = page_size;
-    LOCK_INIT(&iobuf->lock);
+    if (align_flag)
```
not for this patch - we should move to using posix_memalign() instead of all those GF_ALIGN_BUF stuff

mohit84 commented:
/run regression

gluster-ant commented:
1 test(s) failed 
./tests/bugs/gfapi/bug-1319374-THIS-crash.t

0 test(s) generated core 


3 test(s) needed retry 
./tests/000-flaky/basic_afr_split-brain-favorite-child-policy.t
./tests/000-flaky/glusterd-restart-shd-mux.t
./tests/bugs/gfapi/bug-1319374-THIS-crash.t
https://build.gluster.org/job/gh_centos7-regression/1735/

mohit84 commented:
@Saju We can't consider this patch for today release build because bug-1319374-THIS-crash.t is crashing, though crash is not specific to above patch but with this patch the gfapi program is crashing easily during ssl_library loading.Once we will find the correct way to load ssl library then we can consider this patch.

0x00007f19db7c5016 in __strcmp_sse42 () from /lib64/libc.so.6
#1  0x00007f19d9895f09 in getrn () from /lib64/libcrypto.so.10
#2  0x00007f19d98961f0 in lh_insert () from /lib64/libcrypto.so.10
#3  0x00007f19d97e235f in OBJ_NAME_add () from /lib64/libcrypto.so.10
#4  0x00007f19cecca6e9 in SSL_library_init () from /lib64/libssl.so.10
#5  0x00007f19ceef7555 in init_openssl_mt () at socket.c:4125
#6  init (this=0x2f05058) at socket.c:4677
#7  0x00007f19db24cf7e in rpc_transport_load (ctx=ctx@entry=0x2ed84a0, options=options@entry=0x2eff278, 
    trans_name=trans_name@entry=0x2f24408 "gfapi") at rpc-transport.c:345
#8  0x00007f19db250921 in rpc_clnt_connection_init (name=0x2f24408 "gfapi", options=0x2eff278, ctx=0x2ed84a0, clnt=0x2f29258)
    at rpc-clnt.c:1030
#9  rpc_clnt_new (options=options@entry=0x2eff278, owner=<optimized out>, name=name@entry=0x2f24408 "gfapi", 
    reqpool_size=<optimized out>, reqpool_size@entry=8) at rpc-clnt.c:1119
#10 0x00007f19dba61dd3 in glfs_mgmt_init (fs=fs@entry=0x2ed8310) at glfs-mgmt.c:1016
#11 0x00007f19dba5c885 in glfs_volumes_init (fs=fs@entry=0x2ed8310) at glfs.c:260
#12 0x00007f19dba5e249 in glfs_init_common (fs=0x2ed8310) at glfs.c:1093
#13 0x00007f19dba5e36f in pub_glfs_init (fs=0x2ed8310) at glfs.c:1135
#14 0x0000000000400c5e in main (argc=4, argv=0x7ffd1d8ffdb8) at ./tests/bugs/gfapi/bug-1319374.c:114

mykaul commented:
Need to validate which libraries and versions we are using (see https://community.developers.refinitiv.com/questions/12997/the-rfa-application-crashed-in-ssl-library-init-af.html )

mohit84 commented:
> Need to validate which libraries and versions we are using (see https://community.developers.refinitiv.com/questions/12997/the-rfa-application-crashed-in-ssl-library-init-af.html )

The current installed library is (openssl-1.0.2k-16.el7.x86_64), i don't think it is a library issue as we can see symbols are correctly represents on the core dump.

mohit84 commented:
/run full regression

xhernandez commented on libglusterfs/src/iobuf.c:
```diff
@@ -468,79 +441,38 @@ __iobuf_get(struct iobuf_pool *iobuf_pool, const size_t page_size,
-    iobuf->page_size = page_size;
-    LOCK_INIT(&iobuf->lock);
+    if (align_flag)
+        align_size = page_size + GF_IOBUF_ALIGN_SIZE;
```
```suggestion
        align_size = page_size + GF_IOBUF_ALIGN_SIZE - 1;
```

Otherwise, in cases that `page_size` is already aligned to `GF_IOBUF_ALIGN_SIZE`, it will waste one entire page of memory.

mykaul replied:
Certainly one of the properties we've lost in this change is that iobuf->ptr are no longer naturally aligned and we need now more memory than we've used to. This is noticeable in all block sizes:
- When in the past a 4K iobuf would have been just 4K taken from a mmap'ed arena, now we need 8K
- A 128K (which is our default page size, so all iobuf_get() use it) that could have been perfectly aligned as it was taken from an mmap'ed arena, now requires more, so doesn't really fit into tcmalloc's 128K recommendation.

The first means that we can't easily remove (for example) the extra memory copies in __posix_write(), the 2nd means that ```if (page_size <= USE_IOBUF_POOL_IF_SIZE_GREATER_THAN) {``` is probably not optimal. A hack would be convert:
```
if (page_size <= USE_IOBUF_POOL_IF_SIZE_GREATER_THAN) {
```
to:
```
if (page_size < USE_IOBUF_POOL_IF_SIZE_GREATER_THAN) {
```

gluster-ant commented:
1 test(s) failed 
./tests/bugs/gfapi/bug-1319374-THIS-crash.t

0 test(s) generated core 


4 test(s) needed retry 
./tests/000-flaky/glusterd-restart-shd-mux.t
./tests/basic/volume-snapshot-clone.t
./tests/bugs/gfapi/bug-1319374-THIS-crash.t
./tests/bugs/glusterfs/bug-866459.t

1 flaky test(s) marked as success even though they failed 
./tests/000-flaky/glusterd-restart-shd-mux.t

mohit84 committed:
iobuf: Resolve reviewer comments

Fixes: #2850
Signed-off-by: Mohit Agrawal <moagrawa@redhat.com>

mohit84 commented:
/run regression

mykaul commented on libglusterfs/src/iobuf.c:
```diff
@@ -23,8 +23,9 @@
     /* { pagesize, num_pages }, */
-    {128, 1024},     {512, 512},       {2 * 1024, 512}, {8 * 1024, 128},
-    {32 * 1024, 64}, {128 * 1024, 32}, {256 * 1024, 8}, {1 * 1024 * 1024, 2},
+    {132 * 1024, 16},
```
I'm pretty sure that on one hand, 16 are not enough, OTOH, you should have created a pool of 128K as well (which is the default size). I'd go with 64 of 128K and 64 of 132. What I'm unsure is if 132 is enough, or we need 128K+4K+few more bytes...

mykaul commented on libglusterfs/src/iobuf.c:
```diff
@@ -23,8 +23,9 @@
-    {32 * 1024, 64}, {128 * 1024, 32}, {256 * 1024, 8}, {1 * 1024 * 1024, 2},
+    {132 * 1024, 16},
+    {256 * 1024, 8},
+    {1 * 1024 * 1024, 2},
```
I wonder when is it useful to have it. On one hand, there are only 2 of them per pool, otoh, they seem rare enough that the cost of allocation/deallocation isn't that high.
